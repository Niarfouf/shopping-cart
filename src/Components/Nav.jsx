import { Outlet, Link } from 'react-router-dom'
import { useState, useEffect } from 'react'

export default function Nav({ cart }) {
    const [dropDown, setDropDown] = useState('no')

    const [categories, setCategories] = useState([])
    async function fetchCategories() {
        try {
            const response = await fetch(
                'https://fakestoreapi.com/products/categories'
            )
            const data = await response.json()
            setCategories(data)
        } catch (e) {
            console.error(e)
        }
    }
    useEffect(() => {
        fetchCategories()
    }, [])
    function capitalizeFirstLetter(string) {
        return string.charAt(0).toUpperCase() + string.slice(1)
    }
    return (
        <>
            <header className="navbar">
                <nav>
                    <ul>
                        <li>
                            <Link to={``}>Home</Link>
                        </li>
                        <li
                            className="drop-down"
                            onMouseEnter={() => setDropDown('yes')}
                            onMouseLeave={() => setDropDown('no')}
                        >
                            <button className="drop-button">Shop</button>
                            <div
                                onClick={() => setDropDown('no')}
                                className={
                                    dropDown === 'yes'
                                        ? 'drop-content show'
                                        : 'drop-content'
                                }
                            >
                                {categories.map((category, index) => {
                                    return (
                                        <Link
                                            key={index}
                                            to={{
                                                pathname: `shop/${category}`,
                                            }}
                                        >
                                            {capitalizeFirstLetter(category)}
                                        </Link>
                                    )
                                })}
                            </div>
                        </li>
                        <li>
                            <Link
                                to={{
                                    pathname: `cart`,
                                }}
                            >
                                Cart{' '}
                                <sup>
                                    {Object.entries(cart).reduce(
                                        (acc, current) =>
                                            acc + current[1].quantity,
                                        0
                                    )}
                                </sup>
                            </Link>
                        </li>
                    </ul>
                </nav>
            </header>

            <main id="content">
                <Outlet />
            </main>
        </>
    )
}
