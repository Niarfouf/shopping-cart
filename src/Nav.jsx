import { Outlet, Link } from 'react-router-dom'
import { useState, useEffect } from 'react'
import Cart from './Cart'
export default function Nav({ deleteCartItems, cart }) {
    const [cartOpen, setCartOpen] = useState('no')
    const [dropDown, setDropDown] = useState('no')

    const [categories, setCategories] = useState([])
    function handleCart() {
        if (cartOpen === 'no') {
            setCartOpen('yes')
        } else {
            setCartOpen('no')
        }
    }

    useEffect(() => {
        async function fetchCategories() {
            try {
                const response = await fetch(
                    'https://fakestoreapi.com/products/categories'
                )
                const data = await response.json()
                setCategories(data)
            } catch (e) {
                console.error(e)
            }
        }
        fetchCategories()
    }, [])
    function capitalizeFirstLetter(string) {
        return string.charAt(0).toUpperCase() + string.slice(1)
    }
    return (
        <>
            <header className="navbar">
                <nav>
                    <ul>
                        <li>
                            <Link to={``}>Home</Link>
                        </li>
                        <li
                            className="drop-down"
                            onMouseEnter={() => setDropDown('yes')}
                            onMouseLeave={() => setDropDown('no')}
                        >
                            <button className="drop-button">Shop</button>
                            <div
                                onClick={() => setDropDown('no')}
                                className={
                                    dropDown === 'yes'
                                        ? 'drop-content show'
                                        : 'drop-content'
                                }
                            >
                                {categories.map((category, index) => {
                                    return (
                                        <Link
                                            key={index}
                                            to={{
                                                pathname: `shop/${category}`,
                                            }}
                                        >
                                            {capitalizeFirstLetter(category)}
                                        </Link>
                                    )
                                })}
                            </div>
                        </li>
                        <li>
                            <button
                                className="cart-button"
                                onClick={handleCart}
                            >
                                Cart {Object.entries(cart).length}
                            </button>
                        </li>
                    </ul>
                </nav>
            </header>
            <Cart
                cartItems={cart}
                cartInfo={cartOpen}
                handleCart={handleCart}
                deleteCartItems={deleteCartItems}
            />
            <main id="content">
                <Outlet />
            </main>
        </>
    )
}
