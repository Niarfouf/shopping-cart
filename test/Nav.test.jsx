import { vi, beforeAll, afterAll } from 'vitest'
import { render, act, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Nav from '../src/Components/Nav'
import { MemoryRouter } from 'react-router-dom'

const item1 = {
    title: 'testTitle1',
    image: 'https://fakestoreapi.com/img/61U7T1koQqL._AC_SX679_.jpg',
    id: 0,
    price: 5.5,
}
const item2 = {
    title: 'testTitle2',
    image: 'https://fakestoreapi.com/img/71kWymZ+c+L._AC_SX679_.jpg',
    id: 1,
    price: 10,
}
const cart = {
    0: { item: item1, quantity: 1 },
    1: { item: item2, quantity: 2 },
}
const cart2 = {
    0: { item: item1, quantity: 5 },
    1: { item: item2, quantity: 2 },
}

describe('Nav', () => {
    //Spy on the global fetch function
    const fetchSpy = vi.spyOn(window, 'fetch')

    //Run before all the tests
    beforeAll(() => {
        //Mock the return value of the global fetch function
        const mockResolveValue = {
            ok: true,
            json: () =>
                new Promise((resolve) =>
                    resolve([
                        'electronics',
                        'jewelery',
                        "men's clothing",
                        "women's clothing",
                    ])
                ),
        }

        fetchSpy.mockReturnValue(mockResolveValue)
    })
    afterAll(() => {
        fetchSpy.mockRestore()
    })
    it('render snapshot', async () => {
        const { container } = render(
            <MemoryRouter>
                <Nav deleteCartItems={() => {}} cart={cart} />
            </MemoryRouter>
        )
        await waitFor(() => expect(fetchSpy).toHaveBeenCalledTimes(1))

        expect(container).toMatchSnapshot()
    })
    it('render correct links', async () => {
        render(
            <MemoryRouter>
                <Nav deleteCartItems={() => {}} cart={cart} />
            </MemoryRouter>
        )
        await waitFor(() => expect(fetchSpy).toHaveBeenCalledTimes(2))

        const links = screen.getAllByRole('link')
        expect(links.length).toBe(6)
    })
    it('render correct cart number', async () => {
        render(
            <MemoryRouter>
                <Nav deleteCartItems={() => {}} cart={cart2} />
            </MemoryRouter>
        )
        await waitFor(() => expect(fetchSpy).toHaveBeenCalledTimes(3))

        expect(screen.getByText('7')).toBeInTheDocument()
    })
})
