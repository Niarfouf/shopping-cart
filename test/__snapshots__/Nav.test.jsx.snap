// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Nav > render snapshot 1`] = `
<div>
  <header
    class="navbar"
  >
    <nav>
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="drop-down"
        >
          <button
            class="drop-button"
          >
            Shop
          </button>
          <div
            class="drop-content"
          >
            <a
              href="/shop/electronics"
            >
              Electronics
            </a>
            <a
              href="/shop/jewelery"
            >
              Jewelery
            </a>
            <a
              href="/shop/men's clothing"
            >
              Men's clothing
            </a>
            <a
              href="/shop/women's clothing"
            >
              Women's clothing
            </a>
          </div>
        </li>
        <li>
          <a
            href="/cart"
          >
            Cart
             
            <sup>
              3
            </sup>
          </a>
        </li>
      </ul>
    </nav>
  </header>
  <main
    id="content"
  />
</div>
`;
